README.md
# app-basis-data

Method to store/manipulate simple data items in multiple backends yet access them in the same way. This is meant to be more than a key value store but less than a DB. There is no way to 'join' data items together to create more complex data items or any way to perform complex queries. Where possible data updates will be atomic, but this may not always be possible.


The initial intention is not have complex rules to match data so no or/and on searches, should throw an error if the search has more than one match

will not support cursors/iterators/paging into data, you can do that yourself once the data has been found and using a cache

This system will not be and is not expected to be fast, its purpose is to allow flexibility in datastore choice, develop on a filebased local system, and move to which ever storage system is available within your environment when you want to go live

* For DBI implementations then tables will be created as needed, DBIx::Class will be used
to keep things simple, though this has its own performance overheads

* if its redis://server:port then we will connect to that
  https://metacpan.org/pod/Redis

* if its mongo:/server:port;database=dddd;collection=cccc
  https://metacpan.org/pod/MongoDBx::Class

Initial suggested stores

* directory + msgpack or sereal files
* DBI
* mongodb
* redis

## Methods

### add

add new data, will have a uniq ID created, will be timestamped
can have unstructured data

### delete, only with uniq ID

### tagsearch

search for entries matching a single tag

tagname
match 'regexp', '='. 'like', '>=' etc
optional from/to timestamp
optional count

### wildsearch

search all entries to match some data

match 'regexp', '='. 'like', '>=' etc
optional from/to timestamp
optional count

### tagcount, wildcount

Matches the tagsearch and wildsearch, but just returns the number of matching items
not the items themselves

### data

get data for a single uniq ID

### update

update an entry (will add if no uniq ID)

### search

from/to
tag, or '*'
(optional) count
- all items must match
fieldname => { 'regexp' => '/*/, '>=' => 123, 'gte' => 'fred' } 
'*' => { 'regexp' => '/fred/'}
'_*' => 

Changes
0.5
  *
0.4 2014-04-29
  * Fix handling of encoder in sri
  * add tests for sereal encoder
0.3 2014-04-29
  * File based data storage completed with tests
  * new Compare module
0.2 2014-04-11
  * update release, not ready yet
0.1 2014-04-08
  * initial github release

lib/App/Basis/Data.pm
# ABSTRACT: Generalised data store access system

=head1 NAME

App::Basis::Data

=head1 SYNOPSIS

  use App::Basis::Data ;
 
  my $store = App::Basis::Data->new( sri => 'file:///tmp/fred' ) ;

  my $id = $store->add( 'test', {message => 'testing', level => 'debug', counter => 20}) ;
  my $large_count_data = $store->wildsearch( counter => { 'gte' => 100} ) ; # or '>=' for numbers
  my $test_count = $store->count( {_tag => { 'eq' => 'test'}) ;
  my $data = $store->data( $id) ;
  $data->{counter} = 50 ;
  my $new_id = $store->update( $data) ;
  say ( $id == $new_id) ? "Data replaced" : "data created" ;
  $store->delete( $id) ;

=head1 DESCRIPTION

I needed a way to store semi-structured data to a number of systems, DB, redis, file.
This module scratches this itch, it is not intended to be fast or efficient, just useful to me.

With this I can store and search for data in a consistent way irrespective of the system
that the data is store to. Document based stores can just store the data

I do not expect (at the moment) that the data is multi-level, for me a data item is a group of
key-values

    item
    fred    => 'bill' ,
    counter => 12,
    source  => 'localhost'

This will not work 

    fred    => 'bill' ,
    counter => 12,
    source  => 'localhost',
    array   => [ 1,2,3,4],
    hash    => { alf => 'alien', friend => 'mork'}

well it will, it can be stored/retrieved, but we can only search scalar keys and values.

=head1 Todo

DBI, Redis, Mongo datastore modules still need to be written

Consider adding L<Net::Graylog::Client> as a datastore

=head1 AUTHOR

 kevin mulholland

=head1 VERSIONS

v0.1  2014/04/08, initial work

=head1 Notes

DBI does not make use of any particular autoincrementing fields for the
id as the different databases do this in different ways, so we are stuck with 
implementing a settings table and a data table 
 
=head1 See Also


=head1 Todo


=cut

package App::Basis::Data;

use 5.10.0;
use strict;
use warnings;
use POSIX qw(strftime);
use Data::Printer;
use Sys::Hostname;
use Data::UUID;
use App::Basis;
use Try::Tiny;
use Moo;
use Date::Manip;
use namespace::clean;

# -----------------------------------------------------------------------------

=head1 Public Functions

=over 4

=cut

=item new

Create a new instance of the data store

    my $store = App::Basis::Data->new( sri => 'file:///tmp/store') ;

B<Parameters>
  sri storage resource indicator, expected to have file:///, dbi:, redis: and mongo:
  
=cut

has sri       => ( is => 'ro', required => 1 );
has _uuid     => ( is => 'ro', init_arg => undef, default => sub { Data::UUID->new() }, );
has _hostname => ( is => 'ro', init_arg => undef, default => sub { hostname(); } );
has _handler  => ( is => 'ro', init_arg => undef, default => sub { } );
has _module   => ( is => 'ro', init_arg => undef, default => sub { } );

# create a date::manip object so we don't have to keep doing it, useful in long running
# situations
has _datemanip => ( is => 'ro', init_arg => undef, default => sub { Date::Manip::Date->new() } );

# -----------------------------------------------------------------------------
# remove leading and trailing spaces
sub _trim {
    my ($str) = @_;
    return if ( !$str );

    $str =~ s/^\s+//gsm;
    $str =~ s/\s+$//gsm;

    return $str;
}

# -----------------------------------------------------------------------------

# this is where we will initialise _handler with the object that will perform the
# actual data storage functions

sub BUILD {
    my $self = shift;

    my ($module) = ( $self->sri =~ /^(\w+):/ );
    my $class = "App::Basis::Data::" . ucfirst($module);

    $self->{_module} = $module;
    try {
        eval "require $class";
    }
    catch {
        say STDERR "$class vs $_";
        die "No handler for $module";
    };

    # while each SQL DB understands its DBI format, this is not true of the redis/mongo DBs
    # so lets also split out any component parts

    my %params = ( sri => $self->sri );
    foreach my $s ( split( ';', $self->sri ) ) {
        my ( $k, $v ) = ( $s =~ /^(.*?)=(.*)/ );
        if ($k) {
            $k = _trim($k);
            $params{$k} = _trim($v);
        }
    }

    # and create the required class instance
    $self->{_handler} = $class->new(%params);
}

# -----------------------------------------------------------------------------
sub _as_timestamp {
    my $self = shift;
    my ($datestr) = @_;
    my $timestamp;
    my $date = $self->{_datemanip}->new();

    # we may have to convert the timestamp
    if ( $datestr && $datestr !~ /^\d+$/ ) {

        # reuse the date::manip object
        $date->parse($datestr);
        {
            # if we get a warning about converting the date to a day, there
            # must be a problem with parsing the input date string
            local $SIG{__WARN__} = sub {
                die "Invalid date";
            };
            my $day = $date->printf("%a");
        }

        # update timestamp with unix epoch
        $timestamp = $date->secs_since_1970_GMT;
    }
    else {
        $timestamp = $datestr;
    }

    return $timestamp;
}

# -----------------------------------------------------------------------------

=item add 

Add a new data item to the store

    my $store = App::Basis::Data->new( sri => 'file:///tmp/store') ;
    my $id = $store->add( thing1 => 123, thing2 => 'abc') ;

B<Parameters>
  hash of data to store
  timestamp will override automatic timestamp, timestamp only valid from 1970
  
=cut

sub add {
    my $self = shift;
    my ( $tag, $params ) = @_;

    # we don't add empty data
    return 0 if ( !$params );

    die "add requires a hashref" if ( ref($params) ne 'HASH' );

    # remove any parameters that are prefixed '_'
    foreach my $key ( keys %$params ) {
        delete $params->{key} if ( $key =~ /^_/ );
    }

    $params->{_timestamp} = $self->_as_timestamp( $params->{timestamp} || time() );
    $params->{_created}   = time();
    $params->{_source}    = $self->_hostname;
    $params->{_uuid}      = $self->_uuid->create_str();
    $params->{_tag}       = $tag;

    return $self->{_handler}->add( $tag, $params );
}

# -----------------------------------------------------------------------------

=item taglist

list all the tags that have been used to tag data

    my $store = App::Basis::Data->new( sri => 'file:///tmp/store') ;
    my $@tags = $store->taglist() ;
  
=cut

sub taglist {
    my $self = shift;
    my ($tag) = @_;
    $self->{_handler}->taglist($tag);
}

# -----------------------------------------------------------------------------

=item delete

Remove a data item from the store

    my $store = App::Basis::Data->new( sri => 'file:///tmp/store') ;
    my $id = $store->add( thing1 => 123, thing2 => 'abc') ;
    $store->delete( $id) ;
    # alternatively
    my $data = $store->($id) ;
    $store->delete( $data) ;

B<Parameters>
    id or data item

=cut

sub delete {
    my $self = shift;
    my ($id) = @_;

    # we may get passed the full data item
    $id = $id->{_id} if ( ref($id) eq 'HASH' );
    $self->{_handler}->delete($id);
}

# -----------------------------------------------------------------------------

=item data

Get a data item from the store

    my $store = App::Basis::Data->new( sri => 'file:///tmp/store') ;
    my $id = $store->add( thing1 => 123, thing2 => 'abc') ;
    my $data = $store->data( $id) ;

B<Parameters>
  id   valid data item id

B<Returns>
  hashref of data

=cut

sub data {
    my $self = shift;
    my ($id) = @_;

    # we may get passed the full data item
    $id = $id->{_id} if ( ref($id) eq 'HASH' );

    $self->{_handler}->data($id);
}

# -----------------------------------------------------------------------------

=item update

Update a data item

    my $store = App::Basis::Data->new( sri => 'file:///tmp/store') ;
    my $id = $store->add( thing1 => 123, thing2 => 'abc') ;
    my $data = $store->data( $id) ;
    $data->{fred} = 'wilma' ;
    delete $data->{thing2} ;
    my $s = $store->update( $data) ;
    say ($s == $id ? "updated" : "new record added")  ;

B<Parameters>
  hashref of data to update

B<Returns>
  id of updated item
  
=cut

sub update {
    my $self = shift;
    my ($params) = @_;
    die "update requires a hashref" if ( ref($params) ne 'HASH' );
    my $id      = $params->{_id};
    my $current = $self->data($id);

    # basic check to see if its the same
    if ( $current && $current->{_uuid} eq $params->{_uuid} ) {
        $params->{_modified} = time();

        # remove any parameters that are prefixed '_'
        foreach my $key ( keys %$params ) {
            delete $params->{key} if ( $key =~ /^_/ );
        }

        # add in the entries from the current record
        foreach my $key ( keys %$current ) {
            $params->{key} = $current->{key} if ( $key =~ /^_/ );
        }
        $id = $self->{_handler}->update($params) ? $id : undef;
    }
    else {
        $id = $self->add( $params->{_tag}, $params );
    }
    return $id;
}

# -----------------------------------------------------------------------------

=item search

search for matching items
This is the general purpose search

    my $store = App::Basis::Data->new( sri => 'file:///tmp/store') ;
    my $data = $store->search( 
        { 
            _tag => { 'eq' => 'fred'},
            _timestamp = { '>=' => '2013-01-01 12:00:00', '<=' => 'yesterday'},
            thing2 => { '~' => 'a'},
        }
    ) ;

valid comparisons are 
    numbers: >= > = != =<     (=> same as <=, <= same as =<)
    strings: gte gt eq ne lt lte      (ge same as gte, le same as lte)
    time: prefix string compare with 'time:', i.e. time:eq, 
    date: prefix string compare with 'date:', i.e. date:eq (date:after same as date:gt, date:before same as date:lt)
    there are no :ne comparisons for date or time
    regular expressions: ~ and !~ 
    see L<App::Basis::Data::Compare> for more infomation.

B<Parameters>
  hashref of things to search against

B<Returns>
  arrayref of matching items
  
=cut

# implementation note: the handler version of this method should only grab the matching tag items
# and match against any underscore prefixed item, then return the items here so that the more
# complex search can be done in one place and consistently

sub search {
    my $self = shift;
    my ($params) = @_;
    die "search requires a hashref" if ( ref($params) ne 'HASH' );

    my $items = $self->{_handler}->search($params);
}

# -----------------------------------------------------------------------------

=item ssearch

search for matching items using strings
This is the general purpose search

    my $store = App::Basis::Data->new( sri => 'file:///tmp/store') ;
    my $data = $store->ssearch( 
            [
            '_tag  eq fred',
            '_timestamp >= 2013-01-01 12:00:00', 
            '_timestamp <= 2014-01-01 12:00:00',
            'thing2 ~ a',
        ]
    ) ;

valid comparisons are 
    numbers: >= > = != =<     (=> same as <=, <= same as =<)
    strings: gte gt eq ne lt lte      (ge same as gte, le same as lte)
    time: prefix string compare with 'time:', i.e. time:eq, 
    date: prefix string compare with 'date:', i.e. date:eq (date:after same as date:gt, date:before same as date:lt)
    there are no :ne comparisons for date or time
    regular expressions: ~ and !~ 
    see L<App::Basis::Data::Compare> for more infomation.

B<Parameters>
  hashref of things to search against

B<Returns>
  arrayref of matching items
  
=cut

# implementation note: the handler version of this method should only grab the matching tag items
# and match against any underscore prefixed item, then return the items here so that the more
# complex search can be done in one place and consistently

sub ssearch {
    my $self = shift;
    my ($strings) = @_;
    die "ssearch requires an arrayref" if ( ref($strings) ne 'ARRAY' );

    my $params;

    # build up the standard search
    foreach my $s ( @{$strings} ) {
        my ( $field, $op, $cmp ) = ( $s =~ /^(\w+)\s+(.*?)\s+(.*)/ );
        next if ( !$field );

        if ( $params->{$field} ) {
            $params->{$field}->{$op} = $cmp;
        }
        else {
            $params->{$field} = { $op => $cmp };
        }
    }

    # and do the normal search
    my $items = $self->{_handler}->search($params);
}

# -----------------------------------------------------------------------------

=item count

Count the number of items that match a search

    my $store = App::Basis::Data->new( sri => 'file:///tmp/store') ;
    my $items = $store->count( {  
         _tag       => { 'eq'       => 'bill' },
        _timestamp => { 'date:gte' => '2013-01-01 00:00:00', 'date:lte' => '2013-12-31 23:59:59' },
    }) ;

B<Parameters>
  search terms to find items to count
  
=cut

sub count {
    my $self = shift;
    my ($rules) = @_;
    die "count requires a hashref" if ( $rules && ref($rules) ne 'HASH' );

    $self->{_handler}->count($rules);
}

# -----------------------------------------------------------------------------

=item purge items that match a search

Count the number of items with the tag, parameters are optional

    my $store = App::Basis::Data->new( sri => 'file:///tmp/store') ;
    my $items = $store->purge( {  
         _tag       => { 'eq'       => 'bill' },
        _timestamp => { 'date:gte' => '2013-01-01 00:00:00', 'date:lte' => '2013-12-31 23:59:59' },
    }) ;

B<Parameters>
    search terms to find items to purge
  
B<Returns>
    Number of items deleted

=cut

sub purge {
    my $self = shift;
    my ($rules) = @_;
    die "purge requires a hashref" if ( $rules && ref($rules) ne 'HASH' );

    $self->{_handler}->purge($rules);
}

# -----------------------------------------------------------------------------

=back 

=cut

# -----------------------------------------------------------------------------
1;


lib/App/Basis/Data/File.pm
# ABSTRACT: File based data store access system

=head1 NAME

App::Basis::Data::File

=head1 SYNOPSIS

  use App::Basis::Data ;
 
  my $store = App::Basis::Data->new( sri => 'file:///tmp/fred' ) ;

  my $id = $store->add( tag => 'test', message => 'testing', level => 'debug', counter => 20) ;
  my $test_data = $store->tagsearch( tag => 'test', from => '2014-01-01') ;
  my $large_count_data = $store->wildsearch( counter => { 'gte' => 100} ) ; # or '>='
  my $test_count = $store->tagcount(tag => 'test') ;
  my $data = $store->data( $id) ;
  $data->{counter} = 50 ;
  my $new_id = $store->replace( id=> $id, $data) ;
  say ( $id == $new_id) ? "Data replaced" : "data created" ;
  $store->delete( $id) ;

=head1 DESCRIPTION

This is a handler class for App::Basis::Data. 

Store semi-structured data to files (JSON by default, sereal optional), allow full range
of App::Basis::Data methods.
 
=head1 AUTHOR

 kevin mulholland

=head1 Notes

You should not use this class directly, it should be called via L<App::Basis::Data>

=cut

package App::Basis::Data::File;

use 5.10.0;
use strict;
use warnings;
use Data::Printer;
use Try::Tiny;
use Moo;
use Path::Tiny;
use App::Basis::Data::Compare;
use namespace::clean;

# -----------------------------------------------------------------------------

=head1 Public Functions

=over 4

=cut

=item new

Create a file based instance of the data store

    my $store = App::Basis::Data->new( sri => 'file:///tmp/store') ;

B<Parameters>
  sri storage resource indicator, expected to have file:///
  
=cut

has sri       => ( is => 'ro', required => 1 );
has _dir      => ( is => 'ro', init_arg => undef, default => sub { } );
has _settings => ( is => 'ro', init_arg => undef, default => sub { } );

has _encoder => ( is => 'ro', init_arg => undef, default => sub {undef} );
has _decoder => ( is => 'ro', init_arg => undef, default => sub {undef} );
has encoder => ( is => 'ro', default => sub {'json'} );

# -----------------------------------------------------------------------------
# _store
# consistent way to store, even if the extra call slows us down
# we add on the we use the encoder name as
# a file extension to make sure we can have multiple formats in one tree
sub _store {
    my $self = shift;
    my ( $path, $data ) = @_;

    # make the directory for the file to live in
    $path .= ".$self->{encoder}" if ( $path !~ /\.$self->{encoder}$/ );
    path( path($path)->dirname )->mkpath;    # will die if any issues
    path($path)->spew( $self->{_encoder}->encode($data) );
    return -f $path;
}

# -----------------------------------------------------------------------------
# _fetch
# we add on the we use the encoder name as
# a file extension to make sure we can have multiple formats in one tree
sub _fetch {
    my $self = shift;
    my ($path) = @_;
    $path .= ".$self->{encoder}" if ( $path !~ /\.$self->{encoder}$/ );
    return $self->{_decoder}->decode( path($path)->slurp );
}

# -----------------------------------------------------------------------------

# this is where we will initialise _handler with the object that will perform the
# actual data storage functions

sub BUILD {
    my $self = shift;

    my ($mname) = ( __PACKAGE__ =~ /App::Basis::Data::(.*)/ );

    # get the basename of this module, as this is the sri we accept
    $mname = lc($mname);

    my ( $module, $dir ) = ( $self->sri =~ m|^($mname)://(.*)| );
    # remove any potential extra arguments
    $dir =~ s/;.*// ;

    die "__PACKAGE__ expects a $mname:// sri" if ( !$module );

    die "Cannot build a datastore as sri points to a file" if ( -f $dir );

    $self->{encoder} = lc( $self->{encoder} );

    if ( $self->{encoder} eq 'json' ) {
        require JSON;
        $self->{_encoder} = JSON->new->allow_nonref;
        $self->{_decoder} = JSON->new->allow_nonref;
    }
    elsif ( $self->{encoder} eq 'sereal' ) {
        require Sereal;
        $self->{_encoder} = Sereal::Encoder->new();
        $self->{_decoder} = Sereal::Decoder->new();
    }
    else {
        die "Unknown encoder required: $self->{encoder}";
    }

    try {
        path($dir)->mkpath;    # will die if any issues
    }
    catch {
        die "Cannot build a datastore as issues with sri ";
    };

    # set the directory for the store, remove trailing dir markers
    $dir =~ s|/$||;
    $self->{_dir} = $dir;
    $self->_get_settings();
}

# -----------------------------------------------------------------------------
# load in and decode the settings
sub _get_settings {
    my $self = shift;

    try {
        $self->{_settings} = $self->_fetch( $self->{_dir} . '/settings' );
    }
    catch {
        # store the defaults
        $self->_store_settings(undef);
    };
}

# -----------------------------------------------------------------------------
# encode and store the settings
sub _store_settings {
    my $self = shift;

    # set defaults if needed
    $self->{_settings} ||= { next_id => '0' };
    $self->_store( $self->{_dir} . '/settings', $self->{_settings} );
}

# -----------------------------------------------------------------------------
# next_id is stored in the settings, this means that settings will be
# reloaded the next_id value updated and saved again
sub _next_id {
    my $self = shift;

    # reload settings and adjust, not quite atomic but fine for simple use
    $self->_get_settings();
    $self->{_settings}->{next_id}++;
    $self->_store_settings();
    return $self->{_settings}->{next_id};
}

# -----------------------------------------------------------------------------
# build a path to access a data item
# we keep things in directories with up to 100 (00..99) items per directory, with 4 levels
# of directory, should allow us to have 100^4 entries in the store!
# we will make
sub _build_path {
    my $self = shift;
    my ( $id, $tag ) = @_;
    $tag ||= 'default';

    # up to 7 leading 0's
    my $fred = sprintf( "%08d", $id );
    my (@b) = ( $fred =~ /(\d{2})/g );
    my $path = $self->{_dir} . "/$tag/data/" . join( '/', @b );

    return $path;
}

# -----------------------------------------------------------------------------
# find the path for a given ID
sub _find_path {
    my $self = shift;
    my ($id) = @_;
    my $path;

    foreach my $tag ( $self->taglist() ) {

        # append the encoder we are using
        $path = $self->_build_path( $id, $tag ) . "." . $self->{encoder};
        last if ( -f $path );
    }

    return $path;
}

# -----------------------------------------------------------------------------
# store the passed data as is
sub add {
    my $self = shift;
    my ( $tag, $params ) = @_;
    my $id;
    my $path;

    # lets give ourselves a few chances to create the ID if another process
    # has created it just before us
    while (1) {
        $id = $self->_next_id();
        $path = $self->_build_path( $id, $tag );
        if ( !-f $path ) {
            last;
        }
        # skip ID forwards up to 4 steps, + 1 for next time into loop = 5
        foreach( 0 .. int(rand( 3))) {
            $id = $self->_next_id();
        }
    }

    if ( !$path ) {
        warn "could not create path";
        return 0;
    }

    # add the ID into the data
    $params->{_id} = $id;
    return $self->_store( $path, $params ) ? $id : undef;
}

# -----------------------------------------------------------------------------
# get the list of all the tags used
sub taglist {
    my $self = shift;
    my @dirs;

    foreach my $dir ( path( $self->{_dir} )->children() ) {
        push @dirs, path($dir)->basename if ( -d $dir );
    }

    return @dirs;
}

# -----------------------------------------------------------------------------
#  only with uniq ID
sub delete {
    my $self = shift;
    my ($id) = @_;

    my $path = $self->_find_path($id);
    if ($path) {
        unlink $path;
    }
    return -f $path ? 0 : 1;
}

# -----------------------------------------------------------------------------
# get data for a single uniq ID
sub data {
    my $self = shift;
    my ($id) = @_;

    return undef if ( !$id );
    my $path = $self->_find_path($id);

    return $self->_fetch( path($path) );
}

# -----------------------------------------------------------------------------
# update the data referenced (will add if no uniq ID)
sub update {
    my $self = shift;
    my ($params) = @_;

    # error if no _id or _tag, would have to create new ertry

    my $path = $self->_build_path( $params->{_id}, $params->{_tag} );

    return 0 if ( !$path );

    return $self->_store( $path, $params );
}

# -----------------------------------------------------------------------------
# search
# this should only match against the terms that are part of the

sub search {
    my $self = shift;
    my ( $rules, $count_only ) = @_;
    my @results;
    my $tag   = '';
    my $count = 0;

    # we can only directly pick tags if they are an exact match
    if ( $rules->{_tag} && $rules->{_tag}->{eq} ) {
        $tag = "/$rules->{_tag}->{eq}";
    }

    my $path = $self->{_dir} . ( $tag ? "$tag/data" : "" );
    my $iter = path($path)->iterator(
        {   recurse         => 1,
            follow_symlinks => 0,
        }
    );

    # we need to grab all the files, check them for things that match
    while ( my $path = $iter->() ) {

        # we only want the files for our encoding
        next if ( !$path->is_file || $path->stringify !~ /\.$self->{encoder}/ );

        # does it look like a file we should be processing?
        next if ( $path->stringify !~ m|$self->{_dir}/.*?/data/\d+?| );
        my $data = $self->_fetch( $path->stringify );

        # if the data does not have one of our uuids then its not one of ours
        next if ( !$data->{_uuid} );

        if ( compare( $data, $rules ) ) {
            if ($count_only) {
                $count++;
            }
            else {
                push @results, $data;
            }
        }
    }

    # decide which results to return
    return $count if ($count_only);

    return \@results;
}

# -----------------------------------------------------------------------------
# Just get the count of results, not the data
sub count {
    my $self = shift;
    my ($rules) = @_;

    # we need a minimum rule to allow searching to take place
    if ( !$rules ) {
        $rules = { _timestamp => { 'date:after' => 0 } };
    }

    # use the special skip through in search to count the matches
    return $self->search( $rules, 1 );
}

# -----------------------------------------------------------------------------
# Just remove the items
sub purge {
    my $self    = shift;
    my ($rules) = @_;
    my $count   = 0;

    foreach my $item ( @{ $self->search($rules) } ) {
        $count += $self->delete( $item->{_id} );
    }

    return $count;
}

# -----------------------------------------------------------------------------
1;


lib/App/Basis/Data/Mongo.pm
# ABSTRACT: XXXXXX based data store access system

=head1 NAME

App::Basis::Data::Xxxxxx

=head1 SYNOPSIS

  use App::Basis::Data ;
 
  my $store = App::Basis::Data->new( sri => 'xxxxxx:///tmp/fred' ) ;

  my $id = $store->add( tag => 'test', message => 'testing', level => 'debug', counter => 20) ;
  my $test_data = $store->tagsearch( tag => 'test', from => '2014-01-01') ;
  my $large_count_data = $store->wildsearch( counter => { 'gte' => 100} ) ; # or '>='
  my $test_count = $store->tagcount(tag => 'test') ;
  my $data = $store->data( $id) ;
  $data->{counter} = 50 ;
  my $new_id = $store->replace( id=> $id, $data) ;
  say ( $id == $new_id) ? "Data replaced" : "data ccreated" ;
  $store->delete( $id) ;

=head1 DESCRIPTION
 

=head1 AUTHOR

 kevin mulholland

=head1 VERSIONS

v0.1  2014/04/08, initial work

=head1 Notes

 
=head1 See Also


=head1 Todo


=cut

package App::Basis::Data::Xxxxxx;

use strict;
use warnings;
use Data::Printer;
use Try::Tiny;
use Moo;
use Path::Tiny;
use Sereal;
use namespace::clean;

# -----------------------------------------------------------------------------

=head1 Public Functions

=over 4

=cut

=item new

Create a file based instance of the data store

    my $store = App::Basis::Data->new( sri => 'Xxxxxx:///tmp/store') ;

B<Parameters>
  sri storage resource indicator, expected to have Xxxxxx:///
  
=cut

has sri => ( is => 'ro', required => 1 );

# -----------------------------------------------------------------------------

# this is where we will initialise _handler with the object that will perform the
# actual data storage functions

sub BUILD {
    my $self = shift;

    my ($mname) = ( __PACKAGE__ =~ /App::Basis::Data::(.*)/) ;
    # get the basename of this module, as this is the sri we accept
    $mname = lc( $mname) ;

    my ( $module, $dir ) = ( $self->sri =~ m|^($mname)://(.*)| );

    die "__PACKAGE__ expects a $mname:// sri" if ( !$module );


}

# -----------------------------------------------------------------------------
sub add {
    my $self = shift;
    my ($tag, $params) = @_;
}

# -----------------------------------------------------------------------------
#  only with uniq ID
sub delete {
    my $self = shift;
    my ($id) = @_ ;
}

# -----------------------------------------------------------------------------
# get the list of all the tags used
sub taglist {
    my $self = shift;
}

# -----------------------------------------------------------------------------
# search for entries matching a single tag

# tagname
# match 'regexp', '='. 'like', '>=' etc
# optional from/to timestamp
# optional count
sub tagsearch {
    my $self = shift;
    my ($tag, $params) = @_;
}

# -----------------------------------------------------------------------------
# search all entries to match some data

# match 'regexp', '='. 'like', '>=' etc
# optional from/to timestamp
# optional count
sub wildsearch {
    my $self = shift;
}

# -----------------------------------------------------------------------------
# Matches the tagsearch and wildsearch, but just returns the number of matching items
# not the items themselves
sub tagcount {
    my $self = shift;
    my ($tag, $params) = @_;
}

sub wildcount {
    my $self = shift;
}

# -----------------------------------------------------------------------------
# get data for a single uniq ID
sub data {
    my $self = shift;
}

# -----------------------------------------------------------------------------
# update the data referenced (will add if no uniq ID)
sub update {
    my $self = shift;
    my ($data) = @_ ;

    # error if no _id or _tag, would have to create new ertry

    my $path = $self->_build_path( $data->{_id}, $data->{_tag}) ;
    my $alt = $self->_find_path( $data->{id}) ;
    # alt and path should match
}
# -----------------------------------------------------------------------------
1;
